File: src/main/java/ru/timeconqueror/timecore/common/capability/property/CoffeeProperty.java
package ru.timeconqueror.timecore.common.capability.property;

import net.minecraft.nbt.CompoundTag;
import ru.timeconqueror.timecore.api.common.blockentity.SerializationType;
import ru.timeconqueror.timecore.common.capability.property.serializer.IPropertySerializer;

public class CoffeeProperty<T> {
    private final String name;
    private final IPropertySerializer<T> serializer;
    private final boolean externallyMutable;

    private boolean clientDependent = false;
    private boolean shouldBeSynced = false;
    private boolean changed = false;

    private T value;

    public CoffeeProperty(String name, T value, IPropertySerializer<T> serializer) {
        this.name = name;
        this.value = value;
        this.serializer = serializer;
        this.externallyMutable = value instanceof MutableProperty;
    }

    public T get() {
        return value;
    }

    public void set(T value) {
        this.changed = true;
        this.value = value;
    }


    public CoffeeProperty<T> clientDependent() {
        clientDependent = true;
        return this;
    }

    public CoffeeProperty<T> synced() {
        shouldBeSynced = true;
        return this;
    }

    public boolean isClientDependent() {
        return clientDependent;
    }

    public boolean isChanged() {
        if (externallyMutable) {
            return ((MutableProperty) value).isChanged() || changed;
        }
        return changed;
    }

    public void setChanged(boolean changed) {
        if (externallyMutable) {
            ((MutableProperty) value).setChanged(changed);
        }
        this.changed = changed;
    }

    public void serialize(CompoundTag nbt, SerializationType type) {
        if (type == SerializationType.SAVE || shouldBeSynced) {
            serializer.serialize(name, value, nbt);
        }
    }

    public void deserialize(CompoundTag nbt) {
        if (nbt.contains(name)) {
            this.value = serializer.deserialize(name, nbt);
            this.changed = false;
        }
    }

    public void deserialize(CompoundTag nbt, boolean fromClient) {
        if (fromClient == clientDependent) {
            deserialize(nbt);
        }
    }

    // ==============================
    //         Kotlin compat
    // ==============================

    public T getValue(Object thisRef, kotlin.reflect.KProperty<?> property) {
        return get();
    }

    public void setValue(Object thisRef, kotlin.reflect.KProperty<?> property, T value) {
        set(value);
    }
}

File: src/main/java/ru/timeconqueror/timecore/common/capability/property/CoffeeProperty.kt
package ru.timeconqueror.timecore.common.capability.property

import net.minecraft.nbt.CompoundNBT
import ru.timeconqueror.timecore.api.common.tile.SerializationType
import ru.timeconqueror.timecore.common.capability.property.serializer.IPropertySerializer
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

class CoffeeProperty<T>(val name: String, private var value: T, val serializer: IPropertySerializer<T>) :
    ReadWriteProperty<Any, T> {

    private var externalChangable = value is IChangable

    var clientDependent = false
        private set

    private var shouldBeSynced = false

    var changed = false
        get() = externalChangable && (value as IChangable).changed || field
        set(value) {
            if (externalChangable) {
                (this.value as IChangable).changed = value
            }
            field = value
        }

    override operator fun getValue(thisRef: Any, property: KProperty<*>) = get()

    override fun setValue(thisRef: Any, property: KProperty<*>, value: T) = set(value)

    fun get() = value
    fun set(value: T) {
        this.changed = true
        this.value = value
    }

    fun serialize(nbt: CompoundNBT, type: SerializationType) {
        if (type == SerializationType.SAVE || shouldBeSynced) {
            serializer.serialize(name, value, nbt)
        }
    }

    fun deserialize(nbt: CompoundNBT) {
        if (nbt.contains(name)) {
            this.value = serializer.deserialize(name, nbt)
            this.changed = false
        }
    }

    fun deserialize(nbt: CompoundNBT, fromClient: Boolean) {
        if (fromClient == clientDependent) {
            deserialize(nbt)
        }
    }

    fun clientDependent(): CoffeeProperty<T> {
        clientDependent = true
        return this
    }

    fun synced(): CoffeeProperty<T> {
        shouldBeSynced = true
        return this
    }
}